SDP Interop 1.0

jallamsetty1 opened the pull request:
- Unified -> Plan-B conversion is same as before.
- Plan-B --> Unified conversion now is based on the current description set on the peerconnection.
1. If the current description doesn't exist, create a new mlines and a new description.
2. If the current description exists, append m-lines.
3. Check for m-lines with direction set to "inactive" before creating a new m-line. These are the m-lines for the sources that have been removed after processing "source-remove".
4. The mids are numbered from 0 onwards (these are default values in the unified plan for local sdps) so that it is easy to re-use them.

gpolitis commented:
This looks good overall! I'd be curious to know why the tests are failing and maybe fix that

I haven't followed this effort closely and miss the bigger picture here so I have a very basic question for my own education..

Is this PR a step towards a unified plan aware lib-jitsi-meet or the general plan is to keep the current status quo (lib-jitsi-meet is plan-b aware for the most part & use the objectified webrtc api where it makes sense & support for unified plan SDP is mainly through this module).

saghul commented:
LGTM overall too, but I haven't dived into the deep details, I trust you on that one ;-) Can you adjust the tests so they pass?

jallamsetty1 commented:
Thank you for the review guys. I will adjust the tests, wanted to get a thumbs up from you guys first about the approach.
@gpolitis, the general plan is to keep the current status quo, i.e., lib-jitsi-meet will be plan-b aware. I tried changing that around in the beginning and that needed a lot of work in our jingle translation logic plus we don't get any added benefits of making it unified plan aware since we only care about the ssrcs and the rest of the media description is the same for all the m-lines. 

We will be using the objectified webrtc api, i.e., the transceiver api to add/remove/replace tracks on the peerconnection and trigger renegotiations only when the browser fires negotiationneeded.
I am wondering if we should worry about backward compatibility since this is a breaking change. Do you know if there are others who are using this module ? 
I tried to use the cached SDP approach like before but I was running into issues when the SDP is modified before sLD/sRD. This module won't be aware of those changes so I am passing the latest description from TPC.

gpolitis commented:
> @gpolitis, the general plan is to keep the current status quo, i.e., lib-jitsi-meet will be plan-b aware. I tried changing that around in the beginning and that needed a lot of work in our jingle translation logic plus we don't get any added benefits of making it unified plan aware since we only care about the ssrcs and the rest of the media description is the same for all the m-lines.

Thanks for explaining, sounds reasonable.

> I am wondering if we should worry about backward compatibility since this is a breaking change. Do you know if there are others who are using this module ?

I don't know who specifically uses this but I believe it's used as there's been some interest in the past. That said, this is a major version bump so I wouldn't worry too much about backwards compatibility.

> I tried to use the cached SDP approach like before but I was running into issues when the SDP is modified before sLD/sRD. This module won't be aware of those changes so I am passing the latest description from TPC.

That's much cleaner than the cached approach, neat trick!

gpolitis commented:
Since there are common points with the previous code, do you think it'd make sense to move index.js  back to interop.js to make diff tracking easier? Also, maybe add a small summary of your changes in THANKS.markdown. A changelog would be nicer but unfortunately we don't have that now.

gpolitis commented on package.json:
```diff
@@ -19,16 +19,27 @@
+    "eslint-plugin-standard": "^4.0.1",
     "jscoverage": "^0.5.5",
-    "jshint": "2.8.0",
+    "lodash.clonedeep": "^4.5.0",
```
do we really need all these dependencies or are there some that we can prune?

jallamsetty1 replied:
We do need lodash but will see if I can do some pruning.

gpolitis replied:
Thanks Jaya. Since you pruned Browserify and migrated over to ESLint, we can probably also remove jshint and watchify?

jallamsetty1 replied:
Sure, will do that.

jallamsetty1 replied:
Sure, will do that.

gpolitis replied:
That's great, thanks Jaya. Since we booted JSHint we should probably remove stale files like `.jshintignore` and `.jshintrc`. Also `npm run lint` fails currently. The error message suggests renaming `.eslintrc.js` to `.eslintrc.cjs`. Could you please check if that fixes the issue?

jallamsetty1 replied:
Oh, looks like we get the error with node 13.1.0 and not with 12.13.0.

jallamsetty1 commented:
> Since there are common points with the previous code, do you think it'd make sense to move index.js back to interop.js to make diff tracking easier? Also, maybe add a small summary of your changes in THANKS.markdown. A changelog would be nicer but unfortunately we don't have that now.

Okay, let me move it back but I am afraid the diff is still going to be a lot. Sure, I will add a summary of my changes to THANKS.markdown.

gpolitis commented on test/sdp_interop.js:
```diff
@@ -549,203 +479,6 @@ a=rtcp-mux\r\n"
     "Not expected Unified Plan output")
 });
 
-QUnit.test('audioInactiveUnifiedPlan2PlanB', function (assert) {
```
I'm trying to figure out what's broken with these test cases, but I can't see the problem, could you please help me understand? Is it not possible to modify them and keep them?

jallamsetty1 replied:
George those specific testcases didn't make sense to me. The plan-b sdp provided didn't have audio mids in the bundle. This is not possible with the transceiver model since when the initial offer is generated, a receiver will always be present even if there is no source added. So the direction will be "recvonly" and a ssrc will be assigned for the receiver. So we will start off with having both the mids (0, 1) in the bundle line and later we won't be able to take them out.
Also, I tried to modify the 2-way-jtisi and 3-way-jitsi json files but there are far too many changes that I have to do make them pass. I was only able to edit half of them. The main differences between the previous translation logic and this one are
1. m-lines are not generated for 'recvonly' ssrcs in the remote description. It doesn't make sense in the transceiver model since there is always a receiver and an ssrc associated for every sender.
2. a=ssrc with msid attribute is always generated in plan-b to unified plan conversion which wasn't happening before.
3. The order of the a=ssrc lines is maintained which wasn't happening before.

gpolitis replied:
> George those specific testcases didn't make sense to me. The plan-b sdp provided didn't have audio mids in the bundle. This is not possible with the transceiver model since when the initial offer is generated, a receiver will always be present even if there is no source added. So the direction will be "recvonly" and a ssrc will be assigned for the receiver. So we will start off with having both the mids (0, 1) in the bundle line and later we won't be able to take them out.

Jitsi Meet doesn't have such a use case, but is it generally impossible to only have inactive audio tracks in the unified plan? How should the plan-b sdp look like in that case?

> Also, I tried to modify the 2-way-jtisi and 3-way-jitsi json files but there are far too many changes that I have to do make them pass. I was only able to edit half of them.

The playbooks capture two main flows: a complete 2-way call and a complete 3-way call (from call join to media exchange). It's okay if there are a lot of changes or if we need to completely rewrite them, but I think it's important to keep them (and even add more, if possible).

Feel free to change the existing playbook format (that is used to represent a call flow) if you find it cumbersome to work with it.

jallamsetty1 replied:
> > George those specific testcases didn't make sense to me. The plan-b sdp provided didn't have audio mids in the bundle. This is not possible with the transceiver model since when the initial offer is generated, a receiver will always be present even if there is no source added. So the direction will be "recvonly" and a ssrc will be assigned for the receiver. So we will start off with having both the mids (0, 1) in the bundle line and later we won't be able to take them out.
> 
> Jitsi Meet doesn't have such a use case, but is it generally impossible to only have inactive audio tracks in the unified plan? How should the plan-b sdp look like in that case?
> 
It is possible to have inactive audio tracks but the bundle line will have the mids of the inactive audio tracks as they would start as "recvonly" first and then be marked as "inactive" later. I will modify the bundle line and bring it back in.
> > Also, I tried to modify the 2-way-jtisi and 3-way-jitsi json files but there are far too many changes that I have to do make them pass. I was only able to edit half of them.
> 
> The playbooks capture two main flows: a complete 2-way call and a complete 3-way call (from call join to media exchange). It's okay if there are a lot of changes or if we need to completely rewrite them, but I think it's important to keep them (and even add more, if possible).
> 
> Feel free to change the existing playbook format (that is used to represent a call flow) if you find it cumbersome to work with it.

Ok, I will rewrite them if I find it cumbersome to edit each one of those descriptions.

gpolitis replied:
Thanks Jaya.

jallamsetty1 commented:
@gpolitis , can you please take a look one more time ? I have adjusted all the unit tests and they are passing now. Thank you so much for all your help ! Will you be able to publish this to npm once you approve the PR ?

gpolitis commented on test/2-way-jitsi/playbook.json:
```diff
@@ -58,33 +51,14 @@
-                "sessionDescription": {
-                    "type": "answer",
-                    "sdp": "v=0\r\no=mozilla...THIS_IS_SDPARTA-46.0.1 8306377039241639027 0 IN IP4 0.0.0.0\r\ns=-\r\nt=0 0\r\na=fingerprint:sha-256 04:FF:CA:8B:C5:DE:9C:19:2C:A8:0A:10:7E:84:DF:5E:FD:CC:28:B3:15:5D:4B:56:F1:AE:1E:CE:8E:05:52:B9\r\na=ice-options:trickle\r\na=msid-semantic: WMS *\r\na=group:BUNDLE audio-1830172161 video-2050122076 data\r\nm=audio 9 RTP/SAVPF 111\r\nc=IN IP4 0.0.0.0\ ...
+                    "sdp": "v=0\r\no=mozilla...THIS_IS_SDPARTA-74.0 6213535423809421069 0 IN IP4 0.0.0.0\r\ns=-\r\nt=0 0\r\na=fingerprint:sha-256 BF:9B:12:7E:9D:21:81:C2:24:CB:95:56:64:FF:0A:39:67:76:A9:76:7C:4D:C1:BF:41:A3:F3:78:27:2D:8E:07\r\na=ice-options:trickle\r\na=msid-semantic: WMS *\r\na=group:BUNDLE audio video\r\nm=audio 9 RTP/SAVPF 111 126\r\nc=IN IP4 0.0.0.0\r\na=rtpmap:111 opus/4800 ...
```
Here we have a `createAnswer` step without a corresponding `setLocalDescription` step. Is it normal?

jallamsetty1 replied:
In TPC, I have removed the unified->plan-b conversions where they were not necessary. We didn't have to do the conversions between createAnswer and setLocalDescription and that is why we don't see it in the call flow.

gpolitis replied:
I see, thanks for explaining Jaya!

gpolitis commented:
> I have adjusted all the unit tests and they are passing now. Thank you so much for all your help !

Thank you Jaya! I left just a couple of really minor comments.

The trick to pass in the current description was really cool. Do you think can update the README file to reflect this change in the _beyond the basics_ section?

Does it mean that Firefox can now be the call offerer?

> Will you be able to publish this to npm once you approve the PR ?

Sure thing.

jallamsetty1 commented:
> > I have adjusted all the unit tests and they are passing now. Thank you so much for all your help !
> 
> Thank you Jaya! I left just a couple of really minor comments.
> 
> The trick to pass in the current description was really cool. Do you think can update the README file to reflect this change in the _beyond the basics_ section?
> 
> Does it mean that Firefox can now be the call offerer?

**Yes, Firefox can now be the offerer and p2p mode works now for unified plan clients. I have added this info in the README.**

> 
> > Will you be able to publish this to npm once you approve the PR ?
> 
> Sure thing.
